node/node.go

```diff
@@ -363,66 +367,277 @@ func (node *Node) AddPendingReceipts(receipts *types.CXReceiptsProof) {
 		Msg("Got ONE more receipt message")
 }
 
+type withError struct {
+	err     error
+	payload interface{}
+}
+
+var (
+	errNotRightKeySize = errors.New("key received over wire is wrong size")
+	errNoSenderPubKey  = errors.New("no sender public BLS key in message")
+)
+
+func validateShardBoundMessage(
+	ctx context.Context, payload []byte,
+) (*msg_pb.Message, error) {
+	var (
+		m         msg_pb.Message
+		senderKey *bls.PublicKey
+	)
+
+	if err := protobuf.Unmarshal(payload, &m); err != nil {
+		return nil, errors.WithStack(err)
+	}
+	var senderPubKeyViaWire []byte
+	maybeCon, maybeVC := m.GetConsensus(), m.GetViewchange()
+
+	if maybeCon != nil {
+		senderPubKeyViaWire = maybeCon.GetSenderPubkey()
+	} else if maybeVC != nil {
+		senderPubKeyViaWire = maybeVC.GetSenderPubkey()
+	} else {
+		return nil, errors.WithStack(errNoSenderPubKey)
+	}
+
+	if len(senderPubKeyViaWire) != shard.PublicKeySizeInBytes {
+		return nil, errors.WithStack(errNotRightKeySize)
+	}
+
+	key, err := bls_cosi.BytesToBLSPublicKey(
```

Comment:
Please add a TODO here to add singleFlight cache, the deserialization of BLS key is very expensive.

Reply:
done.

Reply:
added

Reply:
I don't see the todo?